### ISSUE TITLE
Question about Fokker-Planck example (Integral constraint)

### ISSUE THREAD

**Issue by vaishnavtv** [SciML/NeuralPDE.jl#279]
Hi, I am playing around with the PINN solver for the Fokker-Planck equation, and I don't understand why the following is a correct way of enforcing the normalization constraint:

```julia
# here we use normalization condition: dx*p(x) ~ 1, in order to get non-zero solution.
eq  = Dx((α*x - β*x^3)*p(x)) ~ (_σ^2/2)*Dxx(p(x))+dx*p(x) - 1.
```
From what I understand, this means that at each x for which the equation is evaluated, we inductively require the solution to be such that p(x)*dx ~ 1, but I don't see how this is equivalent to:

```julia
∫p(x)dx = 1.0
```

Alternatively, I am defining an additional loss function which attempts to find the normalization loss using quadrature:
```julia
function norm_loss_function(phi,θ,p)
    prob = QuadratureProblem(phi, lb, ub, θ);
    normP = solve(prob, HCubatureJL(), reltol = 1e-3, abstol = 1e-3);
    loss = abs((normP)[1] - 1.0);
    return loss
end
```
and providing it as an extra input during discretization:
```julia
discretization = NeuralPDE.PhysicsInformedNN(chain,GridTraining(dx), additional_loss = norm_loss_function);
```
The equation was modified to keep only the PDE:
```julia
eq  = Dx((α*x - β*x^3)*p(x)) ~ (_σ^2/2)*Dxx(p(x))
```

With the rest of the code kept the same, this isn't giving me satisfactory results though. 
![normDuffing2](https://user-images.githubusercontent.com/8902316/112509839-dbdbf600-8d5e-11eb-8ce4-401b32396a7e.png)

Please let me know what y'all think. Any assistance would be appreciated.

**Comment by zoemcc** [comment]
Can you try it with a different type of training other than GridTraining?  I've found Grid to often not give the correct results due to only enforcing the PDE constraints at the grid points and not elsewhere.  Try StochasticTraining with a high number of points (128+) or QuadratureTraining with a high number of maxiters (200+).

**Comment by zoemcc** [comment]
Other possible things to tweak:

Optimization method: ADAM for some amount of iterations to get the solution into the right region and then L-BFGS to fine-tune it in that region, or using ADAM with a decaying learning rate.  You can use the Flux.Optimiser to do this decay neatly using a composition of optimisers: 

```julia
Flux.Optimise.Optimiser(ExpDecay(1), ADAM(0.1))
```

**Comment by vaishnavtv** [comment]
Thanks! Will try those tweaks. 
What do you think about the normalization condition for the Fokker-Planck equation? Am I misunderstanding something, or is 
```julia
p(x)*dx = 1
```
the correct representation of the theoretical constraint in the MTK interface?
```julia
∫p(x)dx = 1.0
```

**Comment by zoemcc** [comment]
I believe you're correct. I didn't write that example but from your analysis it looks like it was an approximation to that constraint given that MTK doesn't support integral equations yet (which is being worked on, stay tuned).

**Comment by zoemcc** [comment]
Your additional loss is a good way of embedding that constraint given that we don't support those in the MTK spec yet.

**Comment by vaishnavtv** [comment]
I see. That's a relief.
If I see some good results  with the tweaks you suggested, I'll update this thread. Thank you for the clarification!

**Comment by KirillZubov** [comment]
@vaishnavtv  yes,  `∫p(x)dx = 1.0` is more correct, `p(x)*dx = 1 ` is some approximation of it. it is needed in order not to get a trivial solution.
It looks like you are in the right direction. Try to use one strategy for additional_loss and main loss function( if you use `GridTraining` so and add loss should have the same way of calc points, if `QuadratureProblem` so then `QuadratureTraining`)
and also you need normalizing coefficient  for add_loss 

something like that
```julia
function norm_loss_function(phi,θ,p)
      function loss(x,θ)
            sum(abs2,phi(x, θ) -1, dims=2)
       end
      prob = QuadratureProblem(loss, lb, ub, θ);
       normP = solve(prob, HCubatureJL(), reltol = 1e-3, abstol = 1e-3);
       loss = abs(normP)[1])/number_point #normalizing coefficient 
      return loss
end
```
